File: backend/app/schemas/gradebook.py
import uuid
from datetime import datetime
from pydantic import BaseModel


class StudentGradeOut(BaseModel):
    submission_id:      uuid.UUID
    assignment_id:      uuid.UUID
    assignment_title:   str
    class_name:         str
    subject_name:       str
    subject_code:       str
    max_marks:          int
    due_date:           datetime
    submitted_at:       datetime
    status:             str
    # AI grade (always shown once graded)
    ai_score:           float | None
    formative_feedback: str | None
    # Final grade (shown only after teacher publishes)
    final_score:        float | None
    teacher_comment:    str | None
    is_published:       bool


class AssignmentStatsOut(BaseModel):
    assignment_id:   uuid.UUID
    title:           str
    subject_name:    str
    subject_code:    str
    max_marks:       int
    due_date:        datetime
    is_published:    bool
    enrolled_count:  int
    submitted_count: int
    graded_count:    int
    finalised_count: int
    published_count: int
    avg_ai_score:    float | None
    avg_final_score: float | None


class ClassGradebookOut(BaseModel):
    class_id:        uuid.UUID
    class_name:      str
    enrolled_count:  int
    assignments:     list[AssignmentStatsOut]

File: frontend/lib/api/gradebook.ts
import { api } from "./client";

export interface StudentGrade {
  submission_id:      string;
  assignment_id:      string;
  assignment_title:   string;
  class_name:         string;
  subject_name:       string;
  subject_code:       string;
  max_marks:          number;
  due_date:           string;
  submitted_at:       string;
  status:             string;
  ai_score:           number | null;
  formative_feedback: string | null;
  final_score:        number | null;
  teacher_comment:    string | null;
  is_published:       boolean;
}

export interface AssignmentStats {
  assignment_id:   string;
  title:           string;
  subject_name:    string;
  subject_code:    string;
  max_marks:       number;
  due_date:        string;
  is_published:    boolean;
  enrolled_count:  number;
  submitted_count: number;
  graded_count:    number;
  finalised_count: number;
  published_count: number;
  avg_ai_score:    number | null;
  avg_final_score: number | null;
}

export interface ClassGradebook {
  class_id:       string;
  class_name:     string;
  enrolled_count: number;
  assignments:    AssignmentStats[];
}

export const gradebookApi = {
  getStudentGrades: (): Promise<StudentGrade[]> =>
    api.get("/gradebook/student"),

  getClassGradebook: (classId: string): Promise<ClassGradebook> =>
    api.get(`/gradebook/class/${classId}`),
};
